akka-stream/src/main/scala/akka/stream/impl/fusing/Fusing.scala

```diff
@@ -257,13 +259,76 @@ private[stream] object Fusing {
       else openGroup
 
     if (m.isAtomic) {
-      if (Debug) log(s"atomic module $m")
-      struct.addModule(m, localGroup, inheritedAttributes, indent)
+      m match {
+        case gm: GraphModule if !async ⇒
+          // need to dissolve previously fused GraphStages to allow further fusion
+          if (Debug) log(s"dissolving graph module ${m.toString.replace("\n", "\n" + indent)}")
```

Comment:
we need to measure, of course, but I think this will be faster because much less traversal work is needed since the module is in “canonical form” already